File: backend/emw_backend/agent_tools.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Any, Dict, Optional

from emw_backend.db import SessionLocal
from emw_backend.models import HostSession
from emw_backend.routes.ws import _router  # in-process router (single-worker assumption)


@dataclass
class ToolError(Exception):
    message: str


def hosts_list(*, uid: str) -> Dict[str, Any]:
    now = int(time.time() * 1000)
    with SessionLocal() as db:
        rows = (
            db.query(HostSession)
            .filter(HostSession.firebase_uid == uid)
            .order_by(HostSession.last_seen_at_ms.desc())
            .limit(200)
            .all()
        )

    hosts = []
    for r in rows:
        try:
            caps = json.loads(r.capabilities_json or "{}")
        except Exception:
            caps = {}
        try:
            status = json.loads(r.status_json or "{}")
        except Exception:
            status = {}

        hosts.append(
            {
                "id": r.id,
                "platform": r.platform,
                "device_name": r.device_name,
                "app_version": r.app_version,
                "capabilities": caps,
                "status": status,
                "created_at_ms": r.created_at_ms,
                "last_seen_at_ms": r.last_seen_at_ms,
                "online": (now - (r.last_seen_at_ms or 0)) < 30_000,
            }
        )

    return {"hosts": hosts, "now_ms": now}


def remote_attach(*, uid: str, hostSessionId: str) -> Dict[str, Any]:
    # v1: best-effort. Forward attach to host.
    ok = _router.forward_to_host(uid=uid, host_session_id=hostSessionId, msg={"type": "host.attach", "hostSessionId": hostSessionId})
    if not ok:
        raise ToolError("host_offline")
    return {"attached": True, "hostSessionId": hostSessionId}


def remote_run_script(*, uid: str, hostSessionId: str, name: str, source: str) -> Dict[str, Any]:
    ok = _router.forward_to_host(
        uid=uid,
        host_session_id=hostSessionId,
        msg={"type": "script.run", "hostSessionId": hostSessionId, "name": name, "source": source},
    )
    if not ok:
        raise ToolError("host_offline")
    # scriptInstanceId will arrive asynchronously via script.started; callers should waitForUi / waitForScriptStarted.
    return {"sent": True}


def remote_wait_for_ui(*, uid: str, hostSessionId: str, minRev: int = 0, timeoutSeconds: float = 10.0) -> Dict[str, Any]:
    snap = _router.wait_for_ui_snapshot(uid, hostSessionId, min_rev=int(minRev or 0), timeout_s=float(timeoutSeconds or 10.0))
    if snap is None:
        return {"timeout": True}
    return {"snapshot": snap}


def remote_send_ui_event(
    *,
    uid: str,
    hostSessionId: str,
    scriptInstanceId: str,
    targetNodeId: str,
    name: str,
    payload: Optional[Dict[str, Any]] = None,
    baseRev: Optional[int] = None,
) -> Dict[str, Any]:
    msg: Dict[str, Any] = {
        "type": "ui.event",
        "hostSessionId": hostSessionId,
        "scriptInstanceId": scriptInstanceId,
        "targetNodeId": targetNodeId,
        "name": name,
        "payload": payload or {},
    }
    if baseRev is not None:
        msg["baseRev"] = int(baseRev)

    ok = _router.forward_to_host(uid=uid, host_session_id=hostSessionId, msg=msg)
    if not ok:
        raise ToolError("host_offline")
    return {"sent": True}

File: web/src/server/agentTools.ts
import type { RemoteIncomingPayload, RemoteSessionState } from "./ws/state";
import { hostSessionsStore } from "./store/hostSessions";

export class ToolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ToolError";
  }
}

export function hostsList(uid: string) {
  return hostSessionsStore.list(uid);
}

export function remoteAttach(remoteState: RemoteSessionState, uid: string, hostSessionId: string) {
  const host = remoteState.getHost(uid, hostSessionId);
  if (!host) throw new ToolError("host_offline");
  remoteState.forwardToHost(uid, hostSessionId, { type: "host.attach", hostSessionId });
  return { attached: true, hostSessionId };
}

export function remoteRunScript(remoteState: RemoteSessionState, uid: string, hostSessionId: string, name: string, source: string) {
  const ok = remoteState.forwardToHost(uid, hostSessionId, {
    type: "script.run",
    hostSessionId,
    name,
    source,
  });
  if (!ok) throw new ToolError("host_offline");
  return { sent: true };
}

export async function remoteWaitForUi(remoteState: RemoteSessionState, uid: string, hostSessionId: string, minRev = 0, timeoutSeconds = 10) {
  const snapshot = await remoteState.waitForUiSnapshot(uid, hostSessionId, minRev, timeoutSeconds * 1000);
  return snapshot ? { snapshot } : { timeout: true };
}

export function remoteSendUiEvent(
  remoteState: RemoteSessionState,
  uid: string,
  payload: {
    hostSessionId: string;
    scriptInstanceId: string;
    targetNodeId: string;
    name: string;
    payload?: Record<string, unknown>;
    baseRev?: number | null;
  },
) {
  const ok = remoteState.forwardToHost(uid, payload.hostSessionId, {
    type: "ui.event",
    ...payload,
    payload: payload.payload || {},
  });
  if (!ok) throw new ToolError("host_offline");
  return { sent: true };
}

export function toolSchemasV1() {
  return [
    {
      type: "function",
      function: {
        name: "hosts_list",
        description: "List the user’s available host sessions (devices/apps).",
        parameters: { type: "object", properties: {}, additionalProperties: false },
      },
    },
    {
      type: "function",
      function: {
        name: "remote_attach",
        description: "Attach as a controller to a host session to enable remote control.",
        parameters: {
          type: "object",
          properties: { hostSessionId: { type: "string" } },
          required: ["hostSessionId"],
          additionalProperties: false,
        },
      },
    },
    {
      type: "function",
      function: {
        name: "remote_run_script",
        description: "Run a script on an attached host. Provide script source.",
        parameters: {
          type: "object",
          properties: {
            hostSessionId: { type: "string" },
            name: { type: "string" },
            source: { type: "string" },
          },
          required: ["hostSessionId", "name", "source"],
          additionalProperties: false,
        },
      },
    },
    {
      type: "function",
      function: {
        name: "remote_wait_for_ui",
        description: "Wait for the latest ui.snapshot from a host (rev >= minRev).",
        parameters: {
          type: "object",
          properties: {
            hostSessionId: { type: "string" },
            minRev: { type: "integer" },
            timeoutSeconds: { type: "number" },
          },
          required: ["hostSessionId"],
          additionalProperties: false,
        },
      },
    },
    {
      type: "function",
      function: {
        name: "remote_send_ui_event",
        description: "Send a semantic UI event (tap/change/submit/select) to a node id.",
        parameters: {
          type: "object",
          properties: {
            hostSessionId: { type: "string" },
            scriptInstanceId: { type: "string" },
            targetNodeId: { type: "string" },
            name: { type: "string" },
            payload: { type: "object" },
            baseRev: { type: "integer" },
          },
          required: ["hostSessionId", "scriptInstanceId", "targetNodeId", "name"],
          additionalProperties: false,
        },
      },
    },
  ];
}
